### PR TITLE
Enforce the coverage ratchet in CI and guard baseline architecture

### PR DESCRIPTION
**Pull request by stevekinney**
## Summary

Two independent fixes, one branch:

**CIN-604**: cinder's 100% coverage ratchet (`packages/components/scripts/check-coverage-ratchet.ts`) was enforced by no CI workflow and had drifted below its own floor. The decision here is to enforce, not relax: the floor stays at 1.0 for both runtime lines and functions, the drift is closed with real tests, and nothing is deleted just to raise a percentage.

- `bun run --filter=@lostgradient/cinder validate` passes on a clean checkout.
- The ratchet now runs on every pull request as part of the `unit-tests` check (the existing package-job aggregator in `.github/workflows/unit-tests.yaml`, also wired into `main-green.yaml`'s workspace-gates job). `unit-tests` is already a required status check on `main`, so no new branch-protection entry was needed.
- Dropping coverage below the floor fails that check. Verified locally: deleting `file-upload-utilities.test.ts` (which covers both a line and a function) made `test:coverage` exit non-zero with the ratchet naming the exact dropped coverage; restoring the file returned it to exit 0. (Exact numbers below.)
- A handful of lines and branches are genuinely unreachable, not just hard to hit, and each is proven from the calling code's own invariants rather than asserted. `resolveTextDirection`'s trailing fallback and `evaluateContainerSizeConstraints`'s compound width+inline-size clause are marked in place with a `cinder-coverage-unreachable:` source comment tied to the exemption mechanism in `check-coverage-ratchet.ts`, not a bare ignore: the first shares a cache with the ancestor-walk loop that runs first, so any hint it could find would already have caused an earlier return; the second's only caller already splits that condition into single-feature clauses before this function ever sees it, since the validator and the evaluator share the same `splitTopLevel`/`unwrapRedundantParentheses` helpers. Two more are proven unreachable the same way but closed by restructuring around the unreachable edge instead of marking it, since each is a plain early exit with no nested function to worry about: `splitScopeOutsideContext`'s dead ternary arm (its mandatory preprocessing step, `normalizeScopeRelativeSelector`, already rules the empty-`before` shape out before this function runs) is collapsed onto its own guard clause; `composedFocusScopes`'s `while` loop tracks its own exit condition in a flag instead of relying on a condition that can never actually be the one that ends the loop (every root node an `Element` anchor's `getRootNode()` chain can produce—Element, Document, ShadowRoot, or DocumentFragment—always has `querySelectorAll`, so `isSearchableRoot` can never be the false operand of the loop's original condition). Separately, a fixture script's `import.meta.main` CLI guard is exempted by filename in `check-coverage-ratchet.ts` itself, the same way `src/cli/index.ts` already was, since it only runs as a subprocess and Bun doesn't merge subprocess LCOV.

Two things worth calling out about how this closed:

Bun's `--coverage-reporter=lcov` output has no per-function `FN:`/`FNDA:` records, only a file's aggregate `FNF`/`FNH` totals, so 100% *lines* does not imply 100% *functions*. Six functions across four files were created but never invoked anywhere, invisible to the line-only diagnostics that caught everything else: `attachments.ts`'s two fallback `cancelAnimationFrame` substitutes, `chart-scale.ts`'s three scale-factory closures, `file-upload-utilities.ts`'s no-owner-window fallback, and the implicit constructor Bun's LCOV synthesizes for a class with private-field initializers in `terminal-output-parser.ts` (fixed with an explicit empty constructor, which needed a scoped `eslint-disable-next-line no-useless-constructor` since it isn't actually useless here—it's the only thing that gets the synthesized constructor credited).

Of the four sites this effort originally called Bun coverage artifacts, empirical isolation (reverting each to its original shape one at a time, in isolation, and checking whether the exact same line stays uncovered with every existing test present) found zero actually were. Two were genuinely untested branches masked by the restructure giving a line a free hit from evaluating a condition, not from the branch's own behavior running: `waitForTransitionCompletion`'s SSR/no-`requestAnimationFrame` fallback (reverted to its original nested `if`/`else`, and given a real test instead—stubs `requestAnimationFrame`/`cancelAnimationFrame` away and proves the listener attaches synchronously; proven to fail without the revert) and `collectComposedElements`'s shadow-slot walk (also reverted to its original nested shape; the missing case was a slot with no assigned nodes at all falling through to its own declared fallback children—untested before this branch, since no existing test exercised an empty, unassigned slot with fallback content. Direct evidence: with the added test present, the closing-brace line this restructure targeted is covered in the reverted nested form exactly as in the flattened form; without that test, the same line under the same nested form is uncovered). The other two are genuinely unreachable edges, not masked branches, and stay restructured with the proof stated in place: `splitScopeOutsideContext`'s dead ternary arm and `composedFocusScopes`'s loop-exit condition (direct evidence for the latter: a detached, parentless element passed as `anchor` still yields a scope rather than nothing, because every root node reachable from an `Element` anchor's `getRootNode()` chain satisfies `isSearchableRoot`, so that condition can never be the one that ends the loop).

`waitForTransitionCompletion`'s hang deserves its own note: under this worktree's installed Bun (1.4.2; the repo's `packageManager` field and CI both pin 1.4.0) with the whole component library preloaded, the masked restructure also caused `bun test` to hang indefinitely partway through the full suite—isolated by reverting one changed source file at a time across four full-suite runs, with an origin/main-content control run completing normally to rule out environmental contention.

One more thing surfaced but out of scope for this fix: `isSearchableRoot`'s type guard signature (`node is Document | ShadowRoot`) is not quite accurate—a detached `Element` also satisfies its runtime check (`'querySelectorAll' in node`), and `composedFocusScopes` can hand one to callers as a scope's `root` in that edge case. Harmless at runtime today (Elements have `querySelectorAll` too), but the type is narrower than what can actually flow through it; left as-is since fixing it is unrelated to the coverage ratchet.

Every coverage number in this effort was measured against Bun 1.4.2 (this worktree's installed version); CI runs the pinned 1.4.0. The Bun-specific instrumentation quirks this branch worked around are emitter behavior, so the first CI run on this branch is also the first toolchain-parity check for them—if it reports a different set of uncovered lines than were seen locally, that's version drift in the instrumentation, not flakiness, and needs the same real-test-or-proven-marker treatment rather than a rerun or a floor change.

**CIN-605**: the baseline Docker wrapper never pinned `--platform`, so on an Apple Silicon host it silently built and ran an arm64 image and rewrote the committed x64 baselines with a different rasterizer. `update-snapshots-docker.ts` now checks host architecture before building at all, and `docker-authenticity.ts` (which already runs inside the container before anything is written) gets a fourth check for container architecture, since the existing three checks all still pass on an arm64 build of the same multi-arch base image. Both failure paths point at the supported route: `gh workflow run browser-tests.yaml -f update_baselines=true -f source_ref=<branch> -f base_ref=main`. `--platform` is deliberately not pinned to force emulation: verifying emulated output is byte-comparable to the committed baselines would mean running the full Playwright suite under emulation and diffing every baseline, which wasn't feasible here. Documented in `docs/visual-regression/baselines.md` along with both guards.

Note: two earlier commit messages on this branch describe conclusions later work superseded. The first coverage commit (`test(components): enforce the coverage ratchet in CI`) names `composedFocusScopes` and `splitScopeOutsideContext` as its Bun-artifact restructures, but its diff also touches `collectComposedElements`'s shadow-slot walk and `waitForTransitionCompletion`, neither mentioned in that message. A later commit (`test(components): cover the slot-fallback path collectComposedElements already handled`) concluded `collectComposedElements` was a genuine Bun artifact; the empirical check above (a test added, then the same line checked covered vs. uncovered across the nested and flattened forms) shows the opposite—it was a masked gap, closed correctly by that commit's added test but for the wrong stated reason. Treat this PR body as the accurate account for all four sites.

## Validation

- `bun run --filter=@lostgradient/cinder test:coverage`: 12329 pass, 6 skip, 0 fail across 483 files (12335 tests); `Coverage ratchet (348 files): functions 100.00% >= 100.00% lines 100.00% >= 100.00%`; exit 0.
- Regression proof (AC3): deleted `file-upload-utilities.test.ts`, ran `test:coverage`, got `Coverage ratchet (348 files): functions 99.93% >= 100.00% lines 99.99% >= 100.00%`, `Failed: runtime functions 99.93% < 100.00%, runtime lines 99.99% < 100.00%.`, `Uncovered runtime lines (1 file(s)): src/components/file-upload/file-upload-utilities.ts: 242`, exit 1. Restored the file, reran, back to `functions 100.00% >= 100.00% lines 100.00% >= 100.00%`, exit 0.
- `bun run --filter=@lostgradient/cinder validate`: exit 0 on a clean checkout (AC1).
- `bun run --filter=@cinder/testing test`: 358 pass, 0 fail; exit 0.
- `bun run --filter=@lostgradient/cinder lint`: exit 0 (0 errors; the one new warning this branch introduced, on `terminal-output-parser.ts`'s explicit constructor, is suppressed with a scoped, documented `eslint-disable-next-line`; no other file this branch touches carries a warning).
- `bun run --filter=@lostgradient/cinder typecheck`: exit 0.

Addresses https://linear.app/lost-gradient/issue/CIN-604 and https://linear.app/lost-gradient/issue/CIN-605.

https://claude.ai/code/session_012d2BKW4ZY57DgfY4xDvL56

### CODE REVIEW

File: .github/workflows/unit-tests.yaml [hunk 1]
```diff
@@ -124,7 +124,7 @@ jobs:
           else
             bunx turbo run test --filter=@lostgradient/markdown --filter=@lostgradient/editor --filter=@lostgradient/cinder-mcp
           fi
-          bunx turbo run test:coverage --filter=@lostgradient/chat $AFFECTED
+          bunx turbo run test:coverage --filter=@lostgradient/chat --filter=@lostgradient/cinder $AFFECTED
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply .github/workflows/unit-tests.yaml]: Coverage and pipeline script changes now select the package lane. Regression cases cover the ratchet and pipeline implementation and test filenames. Verified with the focused suite on Bun 1.4.0. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: packages/components/src/utilities/focus.ts [hunk 2]
```diff
@@ -562,11 +562,22 @@ function isSearchableRoot(node: Node): node is Document | ShadowRoot {
 export function* composedFocusScopes(anchor: Element): Generator<ComposedFocusScope> {
   let referenceNode: Element = anchor;
   let rootNode: Node = anchor.getRootNode();
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/utilities/focus.ts:566]: SearchableRoot now includes Document, DocumentFragment (including ShadowRoot), and Element. The runtime predicate checks for a callable querySelectorAll; a detached-element regression verifies the exported scope contract. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: packages/testing/scripts/docker-authenticity.ts [hunk 3]
```diff
@@ -52,8 +53,29 @@ async function readInstalledPlaywrightVersion(): Promise<string | undefined> {
...
+export function architectureAuthenticityFailure(
+  hostArchitecture: string,
+): AuthenticityFailure | undefined {
+  if (hostArchitecture === REQUIRED_BASELINE_ARCHITECTURE) return undefined;
+  return {
+    check: 'container architecture',
+    expected: REQUIRED_BASELINE_ARCHITECTURE,
+    actual: hostArchitecture,
+  };
+}
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/testing/scripts/docker-authenticity.ts:75]: Renamed the container-side argument to containerArchitecture and corrected the surrounding architecture explanation. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: packages/testing/scripts/update-snapshots-docker.ts [hunk 4]
```diff
@@ -290,6 +327,12 @@ export async function buildPlaywrightDockerImage(
  * forbidden by the plan.
  */
 async function main(): Promise<void> {
+  const architectureGuard = hostArchitectureGuardResult(process.arch);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/testing/scripts/update-snapshots-docker.ts]: The wrapper now reads docker info --format {{.Architecture}} from the daemon, normalizes its native architecture, and rejects unavailable or unsupported results before building. Fake-CLI subprocess regressions verify the exact daemon query, amd64 normalization, and failure handling; container architecture is checked separately. Docker documents the daemon Architecture field at https://docs.docker.com/reference/cli/docker/system/info/. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: packages/testing/scripts/local-bun-version-guard.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { readPinnedBunVersion } from './update-snapshots-docker.ts';
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/testing/scripts/local-bun-version-guard.ts:1]: Added the named check:local-bun-version-guard command, declared its unit-tests/main-green ownership, and invoke it once before test:coverage. Pipeline tests assert its declaration and position; check:pipeline-coverage passes with 83 commands and no warnings. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: packages/components/scripts/check-coverage-ratchet.ts [hunk 6]
```diff
@@ -1,6 +1,10 @@
+import { readFileSync } from 'node:fs';
 import { dirname, isAbsolute, relative, resolve } from 'node:path';
 import { fileURLToPath } from 'node:url';
 
+import { localBunVersionNotice } from '@cinder/testing/scripts/local-bun-version-guard.ts';
+import { readPinnedBunVersion } from '@cinder/testing/scripts/update-snapshots-docker.ts';
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-coverage-ratchet.ts]: The Cinder test:coverage task now hashes packages/testing/scripts/**, covering the cross-workspace guard implementations and their transitive script imports. Added a Turbo input contract assertion. Fixed in 6790ceb7938cfc85ffaa5728466baec09588a8fb.

File: .github/workflows/unit-tests.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevekinney) [reply .github/workflows/unit-tests.yaml]: The final scope audit also found that styles-only changes could skip the package lane. Commit 01e904786 includes that lane for every scope result, with style, documentation, and guard-script regressions. The package job always invokes coverage without --affected, and required unit-tests enforces its result. The branch now includes origin/main through 8b998570c; current head is 21191c539.
